labgrid/util/ssh.py

```diff
@@ -0,0 +1,324 @@
+            host=self.host,
+            command=command,
+        ).split(' ')
+        res = subprocess.check_call(
+            complete_cmd,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL
+        )
+
+        return res
+
+    @check_connected
+    def run_command(self, command):
+        """Run a command over the SSHConnection
+
+        Args:
+            command (string): The command to run
+
+        Returns:
+            int: exitcode of the command
+        """
+        return self._run_command(command)
+
+    @check_connected
+    def get_file(self, remote_file, local_file):
+        """Get a file from the remote host"""
+        subprocess.check_call([
+            "scp", "-o", "ControlPath={}".format(self._socket),
+            "{}:{}".format(self.host, remote_file), "{}".format(local_file)
+        ])
+
+    @check_connected
+    def put_file(self, local_file, remote_path):
+        """Put a file onto the remote host"""
+        subprocess.check_call([
+            "rsync", "-e", "ssh -o ControlPath={}".format(self._socket),
+            "{}".format(local_file), "{}:{}".format(self.host, remote_path)
+        ])
+
+    @check_connected
+    def add_port_forward(self, remote_host, remote_port):
+        """forward command"""
+        local_port = get_free_port()
+        destination = "{}:{}".format(remote_host, remote_port)
+
+        # pylint: disable=not-an-iterable
+        if destination in self._forwards:
+            return self._forwards[destination]
+        self._run_socket_command(
+            "forward", " -L {local}:{destination}".format(
+                local=local_port, destination=destination
+            )
+        )
+        self._forwards[destination] = local_port
+        return local_port
+
+    @check_connected
+    def remove_port_forward(self, remote_host, remote_port):
+        """cancel command"""
+        destination = "{}:{}".format(remote_host, remote_port)
+        local_port = self._forwards.pop(destination, None)
+
+        # pylint: disable=not-an-iterable
+        if local_port is None:
+            raise ForwardError("Forward does not exist")
+
+        self._run_socket_command(
+            "cancel", " -L {local}:localhost:{remote}".format(
+                local=local_port, remote=remote_port
+            )
+        )
+
+    def connect(self):
+        self._open_connection()
+        self._connected = True
+
+    @check_connected
+    def disconnect(self):
+        self._disconnect()
+
+    def isconnected(self):
+        return self._connected
+
+    def _check_master(self):
+        args = ["ssh", "-O", "check", "{}".format(self.host)]
+        check = subprocess.call(
+            args,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL
+        )
+        if check == 0:
+            return ""
+
+        return self._start_own_master()
```

Comment:
It seems strange that a check function also starts the master?

Reply:
Since the check function returns the used socket, we can also start the master here if we don't find a socket. The same logic is used in the SSHDriver, I can rework this to be more explicit during the connection process.

Reply:
Or rename the check function to something else that better fit's what it's actually doing. :)